src/sage/schemes/elliptic_curves/kummer_line.py

```diff
@@ -0,0 +1,1765 @@
+# (at your option) any later version.
+#                  https://www.gnu.org/licenses/
+# ****************************************************************************
+
+from sage.misc.cachefunc import cached_method
+
+from sage.rings.integer import Integer
+from sage.schemes.elliptic_curves.ell_generic import EllipticCurve_generic
+from sage.schemes.elliptic_curves.constructor import EllipticCurve
+from sage.schemes.elliptic_curves.ell_point import EllipticCurvePoint_field
+from sage.structure.element import RingElement
+from sage.structure.sage_object import SageObject
+
+
+def KummerLine(curve):
+    r"""
+    Wrapper to construct a Kummer line from an elliptic curve depending on its shape.
+
+    INPUT: ``curve`` -- an elliptic curve in short Weierstrass or Montgomery form
+
+    EXAMPLES::
+
+        sage: E = EllipticCurve(GF(101), [2, 3])
+        sage: K = KummerLine(E); type(K)
+        <class 'sage.schemes.elliptic_curves.kummer_line.KummerLine_weierstrass'>
+
+    This can also deal with Montgomery curves::
+
+        sage: E = EllipticCurve(GF(101), [0, 3, 0, 1, 0])
+        sage: K = KummerLine(E); type(K)
+        <class 'sage.schemes.elliptic_curves.kummer_line.KummerLine_montgomery'>
+
+    Raise an error if not short Weierstrass nor Montgomery::
+
+        sage: E = EllipticCurve(GF(101), [1, 1, 2, 3, 5])
+        sage: K = KummerLine(E)
+        ValueError: curve must be in short Weierstrass or Montgomery form
+    """
+    if not isinstance(curve, EllipticCurve_generic):
+        raise TypeError(
+            "input must be an elliptic curve in short Weierstrass or Montgomery form"
+        )
+
+    ainvs = curve.a_invariants()
+    a, b = ainvs[3], ainvs[4]  # Short Weierstrass
+    AM = ainvs[1]  # Montgomery
+
+    if ainvs == (0, 0, 0, a, b):
+        return KummerLine_weierstrass(curve)
+
+    elif ainvs == (0, AM, 0, 1, 0):
+        return KummerLine_montgomery(curve)
+
+    else:
+        raise ValueError("curve must be in short Weierstrass or Montgomery form")
+
+
+class KummerLine_generic(SageObject):
+    r"""
+    Kummer line class for an elliptic curve.
+
+    EXAMPLES::
+
+        sage: E = EllipticCurve(GF(101), [2, 3])
+        sage: K = KummerLine(E); K
+        Kummer Line of Elliptic Curve defined by y^2 = x^3 + 2*x + 3 over Finite Field of size 101
+    """
+
+    def __init__(self, curve):
+        r"""
+        Constructor for a Kummer line from an elliptic curve.
+
+        INPUT: ``curve`` -- an elliptic curve
+
+        EXAMPLES::
+
+            sage: E = EllipticCurve(GF(101), [2, 3])
+            sage: K = KummerLine(E); K
+            Kummer Line of Elliptic Curve defined by y^2 = x^3 + 2*x + 3 over Finite Field of size 101
+        """
+        self._curve = curve
+        self._base_ring = curve.base_ring()
+
+    def __eq__(self, other):
+        r"""
+        Test equality of Kummer lines by checking if the above curves are the same.
```

Comment:
```suggestion
        Test equality of Kummer lines by checking if the associated curves are the same.
```
Not sure if "associated" is correct, but I think "above" is not.

Reply:
(An elliptic curve lies "above" its Kummer line in the sense that it double-covers it; the choice of word is not uncommon.)

Reply:
Ahh I misinterpreted as the line above (`self` and `other`) 🥲 my bad